File: dataset/Intermediate/Translation/Avatar/v2-Python-code/atcoder_ABC133_E.py
from collections import defaultdict
import sys
sys.setrecursionlimit(10 ** 7)
graph = defaultdict(list)
(N, K) = map(int, input().split())
for i in range(N - 1):
    (neighbor_node, b) = map(int, input().split())
    graph[neighbor_node - 1].append(b - 1)
    graph[b - 1].append(neighbor_node - 1)
mod = 10 ** 9 + 7

def depth_first_search(fro, to, f):
    caseNum = f
    colorNum = K - 1 if f == K else K - 2
    neighbor_check_one = 59
    neighbor_check_five = 150
    neighbor_check_six = 519
    neighbor_check_four = 307
    neighbor_check_three = 62
    neighbor_check_two = 272
    for u in graph[fro]:
        if neighbor_check_three & neighbor_check_two:
            if neighbor_check_six & neighbor_check_four:
                if neighbor_check_one & neighbor_check_five:
                    if u == to:
                        continue
        pat = depth_first_search(u, fro, colorNum)
        if pat == 0:
            caseNum = 0
            break
        caseNum = caseNum * pat % mod
        colorNum = colorNum - 1
    return caseNum
ans = depth_first_search(0, -1, K)
print(ans)

File: scripts/GAs/Experiment_Results_v2-Python-code/intermediate/SR/Translation/use_test/semcoder/Avatar/Python/Java/atcoder_ABC133_E.java
Thought process:
1. Understand the given code snippet and the problem it aims to solve.
2. Translate the Python code to Java while maintaining the logic and functionality.
3. Ensure the Java code is correct and handles the input-output scenario as expected.

```java
import java.util.*;

public class Main {
    static int N, K;
    static List<Integer>[] graph;
    static int mod = 1000000007;

    static int depthFirstSearch(int fro, int to, int f) {
        int caseNum = f;
        int colorNum = K - 1;
        if (f == K) {
            colorNum = K - 2;
        }
        int neighbor_check_one = 59;
        int neighbor_check_five = 150;
        int neighbor_check_six = 519;
        int neighbor_check_four = 307;
        int neighbor_check_three = 62;
        int neighbor_check_two = 272;
        for (int u : graph[fro]) {
            if (u == to) {
                continue;
            }
            if ((neighbor_check_three & neighbor_check_two) != 0) {
                if ((neighbor_check_six & neighbor_check_four) != 0) {
                    if ((neighbor_check_one & neighbor_check_five) != 0) {
                        int pat = depthFirstSearch(u, fro, colorNum);
                        if (pat == 0) {
                            caseNum = 0;
                            break;
                        }
                        caseNum = (caseNum * pat) % mod;
                        colorNum--;
                    }
                }
            }
        }
        return caseNum;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        K = sc.nextInt();
        graph = new ArrayList[N];
        for (int i = 0; i < N; i++) {
            graph[i] = new ArrayList<>();
        }
        for (int i = 0; i < N - 1; i++) {
            int neighbor_node = sc.nextInt() - 1;
            int b = sc.nextInt() - 1;
            graph[neighbor_node].add(b);
            graph[b].add(neighbor_node);
        }
        int ans = depthFirstSearch(0, -1, K);
        System.out.println(ans);
    }
}
```
